### PR TITLE
Joint-JS state machine animation

### PR DESCRIPTION
**Pull request by szokolai-mate**
Main ticket: #576

The browser client now animates active states of state machines.

Usage:

- Generate JointJS diagram
- Open the diagram in browser
- Launch a tester class. The diagram should animate.

The browser runs a polling loop in JavaScript, querying the `http://localhost:5252/activeElements` URL, which is served by a `com.sun.net.httpserver.HttpServer` wrapped in `hu.elte.txtuml.api.model.execution.diagnostics.DiagnosticsService`

While the diagrams are currently limited to displaying only one instance of each model, the diagnostics data supplied has unique identifiers.

### CODE REVIEW

**Comment by szokolai-mate** [comment]
Jenkins, retest this, please.

File: dev/plugins/hu.elte.txtuml.api.model.execution/src/hu/elte/txtuml/api/model/execution/diagnostics/DiagnosticsServer.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    public void handle(HttpExchange t) throws IOException {
+		//Set required headers
+		Headers headers = t.getResponseHeaders();
+		headers.add("Content-type", "application/json");
+		headers.add("Access-Control-Allow-Origin", "*");
+		
+		//Build the payload
+        String response = "[";
+        Boolean first = true;
+        for( DiagnosticsRegistryEntry e : this.registry.getRegistry()){
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.api.model.execution/src/hu/elte/txtuml/api/model/execution/diagnostics/DiagnosticsServer.java]: Should be simplified, see Java Streams.

File: dev/plugins/hu.elte.txtuml.api.model.execution/src/hu/elte/txtuml/api/model/execution/diagnostics/protocol/GlobalSettings.java [hunk 2]
```diff
@@ -9,4 +9,11 @@
 	 * Sets the VM property key which determines the DiagnosticsPlugin port
 	 */
 	public static final String TXTUML_DIAGNOSTICS_PORT_KEY = "txtUMLDiagnosticsPort";
+	
+	/**
+	 * Parameters for service over HTTP
+	 * NOT KEYS, only predefined parameters.
+	 */
+	public static final Integer TXTUML_DIAGNOSTICS_HTTP_PORT = 52525;
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.api.model.execution/src/hu/elte/txtuml/api/model/execution/diagnostics/protocol/GlobalSettings.java]: We should make the HTTP port configurable from the active txtUML run configuration.

File: dev/plugins/hu.elte.txtuml.export.javascript/src/hu/elte/txtuml/export/javascript/resources/res/globals.js [hunk 3]
```diff
@@ -1,2 +1,8 @@
 //global variables for interscript communication
 var panEnabled = true;
+
+//TODO: generate from eclipse preferences
+//diagnostics constants
+const DIAGNOSTICS_PORT = 52525;
+const DIAGNOSTICS_PATH = "activeElements";
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.javascript/src/hu/elte/txtuml/export/javascript/resources/res/globals.js]: Generate this from `GlobalSettings`.

File: dev/plugins/hu.elte.txtuml.export.javascript/src/hu/elte/txtuml/export/javascript/resources/res/main.css [hunk 4]
```diff
@@ -431,3 +495,42 @@ body {
 		top:0;
 	}
 }*/
+
+/* JointJS animation*/
+
+@keyframes blinking-body {
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.javascript/src/hu/elte/txtuml/export/javascript/resources/res/main.css]: Eclipse cannot parse this.

File: dev/plugins/hu.elte.txtuml.api.model.execution/src/hu/elte/txtuml/api/model/execution/diagnostics/DiagnosticsRegistry.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import hu.elte.txtuml.api.model.ModelClass;
+import hu.elte.txtuml.api.model.StateMachine.Vertex;
+
+/**
+ * Class for storing the state of an execution's diagnostics.
+ * This state consists of active state-machine vertices assigned to instances of state-machine classes.
+ * 
+ * @author szokolai-mate
+ */
+public class DiagnosticsRegistry {
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.api.model.execution/src/hu/elte/txtuml/api/model/execution/diagnostics/DiagnosticsRegistry.java]: Redundant and unsafe implementation, should be replaced by an existing thread safe map.

File: dev/plugins/hu.elte.txtuml.api.model.execution/src/hu/elte/txtuml/api/model/execution/diagnostics/DiagnosticsService.java [hunk 6]
```diff
@@ -94,12 +108,15 @@ public void usingTransition(ModelClass object, Transition transition) {
 
 	@Override
 	public void enteringVertex(ModelClass object, Vertex vertex) {
+		this.registry.addEntry(object, vertex);
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.api.model.execution/src/hu/elte/txtuml/api/model/execution/diagnostics/DiagnosticsService.java]: We should also animate transitions.

**Pull request merged**